qiskit/circuit/parameter.py

```diff
@@ -144,20 +140,27 @@ def __eq__(self, other):
         else:
             return False
 
+    def _hash_key(self):
+        # This isn't the entirety of the object that's passed to `hash`, just the "key" part of
+        # individual parameters.  The hash of a full `ParameterExpression` needs to depend on the
+        # "keys" of `Parameter`s, and our hash needs to be computable before we can be fully
+        # initialised as a `ParameterExpression`, so we break the cycle by making our "key"
+        # accessible separately.
```

Comment:
Actually this was among the very last things I wrote, but I'm fine to change it - I was kind of meaning "initialised" in a general sense rather than specifically a call to `super().__init__`.

I changed the wording to something else in 44711b4c98 - let me know if that's better.

Reply:
Looks reasonable to me.